MAINT: Split lapack_lite more logically across files

eric-wieser opened the pull request:
### Primary goals
 * Remove manually edited files to make upgrading lapack easier (#8376)
 * Move lapack functions into appropriately-named and smaller files
 * Add `<lapack>/INSTALL` to the search path, since some fortran files are duplicated to here.

#### Old structure
 * `dlamch.c` (__manually generated, edited__) - The `dlamch` function. Includes the patch in cadbb5f2ec84cb774023bbe8cbe6f39a93e837d7<sup>1</sup>, so no longer auto-generated
 * `zlapack_lite.c`  - Functions for the `complex128` type. 
 * `dlapack_lite.c` - Every other lapack function<sup>2</sup>

#### New structure
 * `f2c_z_lapack.c` - Functions for the `complex128` (`z`) type.  
 * `f2c_c_lapack.c` - Functions for the `complex64` (`c`) type.
 * `f2c_d_lapack.c` - Functions for the `float64` (`d`) type.  
 * `f2c_s_lapack.c` - Functions for the `float32` (`s`) type.
 * `f2c_config.c` - `dlamch`, `slamch` (for which patch
 * `f2c_lapack.c` - Every other lapack function.

The patch from cadbb5f2ec84cb774023bbe8cbe6f39a93e837d7, which was forgotten, and any other function that comes from the `/INSTALL` directory of lapack<sup>3</sup> is included in `f2c_config.c`. The `make_lite.py` script now applies the `f2c_config.c.patch` to the result, which means that this patch can't be forgotten in future
 


---
<sup>1</sup> Which unfortunately has no description of the bug it was trying to fix

<sup>2</sup> Probably due to the build script not being correctly updated when the `c` and `s` variants were added

<sup>3</sup> (Identical copies of) `[sd]lamch.f` can be found in both `/INSTALL` and `/SRC` within the LAPACK tarball. Previously, we were loading it from `SRC`.  However, in later releases of LAPACK it exists exclusively in `/INSTALL`.  This adds `INSTALL` first on the function lookup path, and puts all the generated code from that path in `misc_lite.c`. This makes the approach forwards-compatible.

## Questions

1. Is `[sd]lamch` special enough to get its own file like this, or can that go in the main `[sd]lapack_.lite.c`
2. Is `slamch` less special than `dlamch` as indicated by the lack of manual generation, or is this a mistake made long ago
3. Is there a better set of names for `lapack_lite.c` and `misc_lite.c` (or single name if they can be merged)

eric-wieser commented:
It's worth pointing out that none of the `.c` files were modified by hand here - these were generated with the invocation of `make_lite.py` in the `README`.

eric-wieser commented:
Hold up, need to reapply cadbb5f2ec84cb774023bbe8cbe6f39a93e837d7

eric-wieser commented:
@charris, can I ping you on this? Thought it would be wise to do some further cleanup before getting #8649 going

charris commented:
Can you document up top what you did in more detail? There are are a bunch of new files, what is in them? The `misc_*` file names are not descriptive. AFAICT, we have the four supported types `c*`, `d*`, `s*`, `d*`, plus supporting routines, it would be nice to have it documented which is where. The `[sd]dlamch` functions are rather special as they determine info we already know for the build, but it probably isn't worth worrying about.

eric-wieser commented:
> The [sd]dlamch functions are rather special

Perhaps, but only `dlamch` was treated specially before, not `slamch`. I notice that there's something going on with `HAVE_CONFIG`, but I can't work out where that's coming from

charris commented:
Also, the current files are divided into complex and real, any reason not to just rename the files `complex_lapack_lite` and `real_lapack_lite`?

We probably need to revise the HAVE_CONFIG stuff, IIRC, it just treats long double, which does not apply here.

eric-wieser commented:
@charris: I think having two 30k line files is probably more manageable than one 60k line file, so I'd argue for `complex64_lapack_lite.c` and `complex128_lapack_lite.c` etc. I'd argue we should keep the names of those four files though, as they emphasize which function names they contain (and ultimately, the heuristic is just "what does the function start with", nothing more.

`misc_lite.c` and `lapack_lite.c` could definitely do with a rename, on the other hand

charris commented:
I'd be happy with four `[s,d,c,z]lapack_lite.c` files. Still, the description up top of this PR should explain "why" and "what" and lay out the new organization. The diffs are huge and I'd rather this only happened once, so is there a reason not to do this in the lapack upgrade PR?

eric-wieser commented:
> I'd be happy with four [s,d,c,z]lapack_lite.c files

Do you want `[ds]lamch` to be folded into these, or do you think they're special enough to deserve their own file?

And what about the stuff that doesn't fit into any of those categories?

> The diffs are huge and I'd rather this only happened once

I'm hoping that the diffs should be smaller when the only changes are actually an upgrade, rather than moving lots of content around. But I guess if you're happy with the 4-way split, I can go ahead locally and verify that the resulting diff is indeed small.

> Still, the description up top of this PR should explain "why" and "what" and lay out the new organization

I'll give this another pass tomorrow. Thanks for the feedback!

eric-wieser commented:
Ok, hoping the up-top description is more useful now

eric-wieser commented on numpy/linalg/lapack_lite/misc_lite.c.patch:
```diff
@@ -0,0 +1,20 @@
+ {
+     /* System generated locals */
+-    real ret_val;
++    volatile real ret_val;
```
This was forgotten in the previous version, so was presumably causing the same bug as the code that the other part of this patch fixed.

charris commented:
This looks much nicer. I guess the only question is the naming/merging of `lapack_lite.c` and `misc_lite.c` files. As the *lamch functions are used in both the blas and lapack files and establish the parameters of the floating type, I'd be inclined to name it something like `mach_lite`,  `machine_lite`, or `floating_lite`.  Or even `install_lite`, although that seems misleading. OTOH, merging the two files seems reasonable also.

I can see that I'm not much help here ;) Do you have any preferences?

charris commented:
Maybe `environment_lite.c`?

charris commented:
I'm coming around to just renaming `misc_lite.c` to `install_lite.c`.

eric-wieser commented:
Perhaps `config_lite.c`? I think I like `environment_lite.c` as well. `install_lite.c` is better than `misc_lite.c` too. 
I'll let you pick one.

charris commented:
`config_lite.c` Sounds good to me. With that name, I think it would also make sense to combine the two files as well,

eric-wieser commented:
> With that name, I think it would also make sense to combine the two files as well

I don't think I agree. `config` because it seems like those are methods you're supposed to tweak based on your system settings.

The current `lapack_lite.c` contains utility functions for lapack that are type-independent. If these are thrown in the same bucket, I don't think this bucket should be called `config`

charris commented:
OK, I don't feel strongly about it.

eric-wieser commented:
Also, is the `_lite` suffix buying us anything at all here?

charris commented:
> Also, is the _lite suffix buying us anything at all here?

Read my mind. The only reason for keeping it I could think of was that it isn't the complete lapack, but rather a "lite" version, as in Miller lite (TM) beer, so the suffix does make some sense.

eric-wieser commented:
What I meant by that is is it needed on each file? I'm all for the folder having that name.

charris commented:
Sure, go ahead and remove the "lite" suffix, I don't have strong feelings either way.

eric-wieser committed:
MAINT: Prefer to load functions from lapack/install if possible

Previously, the dlamch function was a manually-edited file, precluding
autogeneration. We fix this by putting the manual edits in a patch file, and
then generating f2c_config.c from install/*.f, which includes dlamch.

While these functions exist in more than once place in lapack 3.0.0, they only
exist in lapack/install in newer versions. A side effect of this is that some
functions have been pulled out of `f2c_blas.c` and `f2c_s_lapack.c` into this
new file.

The edits in the patch were introduced in cadbb5f2ec84cb774023bbe8cbe6f39a93e837d7,
for a reason that is lost to time.

eric-wieser committed:
BUG: Apply fix for unknown bug in dlamch to slamch too

This bug is mentioned, fixed, but not described, in cadbb5f2ec84cb774023bbe8cbe6f39a93e837d7

To be safe, we patch it for the equivalent float32 function.

eric-wieser commented:
@charris: Done, and rebuilt the commits in slightly more logical order, and to emphasize the bug fix. All intermediate commits build, I think

charris commented:
Oh, now, that is a step too far ;) I do prefer the `[sdcz]lapack.c` names as they follow the lapack conventions.

eric-wieser commented:
> follow the lapack conventions.

If this is something we care about, then we need to only use 6 characters in the name too though, right? ;)

Seriously though - do you want me to revert that, or not?

charris commented:
I really do prefer the sdcz prefixes, as I instantly know what they mean, i.e., functions beginning with the same prefixes.

eric-wieser commented:
How about `lapack_zfuncs`, which also meets the 6 letter rule? I really want to avoid `clapack` being one of the names, because we use that with a different meaning (lapack converted to c code, a now-dead project our code is inspired by).

charris commented:
Geez.  How about `c_` etc prefixes? I doubt anyone will be confused given that all the types are present and we managed with `clapack_lite` all these years. I still prefer the straight forward versions. Next you will be telling me `clap` is a social disease...

eric-wieser commented:
Not as happy with `c_lapack` over `lapack_c` just from a sorting of filenames point of view - which emphasizes that there's one for each type. But I think that's enough to disambiguate `c_lapack.c` from `clapackscrub.py`.

I'll defer to you decision here, but I'd rather not rebase this more than once more. So can you confirm you're definitely sufficiently content with `c_lapack.c`, and would much prefer it to `lapack_c.c`, before I do the rebase?

On a more positive note, I have #8649 working successfully on top of this now

charris commented:
Yeah, I thought if the sorting issue and decided that with so few files it didn't matter much. You could always prepend `_`  to the generated files ;) I also thought `lapack_c.c` was a bit awkward with the two c's. That said, no one is going to be looking at these files again unless something serious goes wrong. I'm happy with the prefixed form.

eric-wieser commented:
> I also thought lapack_c.c was a bit awkward

Yeah, I agree with that

> You could always prepend _ to the generated files

Actually, i think doing something like this is probably a good idea, to discourage editing. How do you feel about `gen_`?

charris commented:
`gen_` sounds good.

charris commented:
Although I think a straight `_` might be better. The prefix `gen_` is a bit confusing as to whether it is a verb or adjective.

eric-wieser commented:
Let's make a decision:

* [ ] `_c_lapack.c` for sorting only
* [ ] `gen_c_lapack.c` for generated
* [ ] `f_c_lapack.c` for fortran
* [ ] `f2c_c_lapack.c` to indicate it came from f2c
* [ ] `c_lapack.c`

charris commented:
I like `f2c_` also.

eric-wieser committed:
MAINT: Improve handing of files and subprocesses

Since we can use 2.7+ features now, we can have the with statement and
subprocess.check_call

eric-wieser committed:
MAINT: Split up the lapack_lite files more sensibly

Also uses this splitting as an excuse to ditch the _lite suffix, in
favor of a f2c_ prefix for all generated files.

Before:

 * `zlapack_lite.c` - Functions for the `complex128` type.
 * `dlapack_lite.c` - Every other lapack function

After:

 * `f2c_z_lapack.c` - Functions for the `complex128` type.
 * `f2c_c_lapack.c` - Functions for the `complex64` type.
 * `f2c_d_lapack.c` - Functions for the `float64` type.
 * `f2c_s_lapack.c` - Functions for the `float32` type.
 * `f2c_lapack.c`   - Every other lapack function

eric-wieser commented:
@charris: All done. `f2c_lite.c` moved to `f2c.c`, so that the rule remains _"`f2c_` prefix means generated"_,

Also took the chance to throw a little bit more cleanup into an extra first commit.

eric-wieser commented on numpy/linalg/lapack_lite/make_lite.py:
```diff
@@ -188,64 +201,61 @@ def getLapackRoutines(wrapped_routines, ignores, lapack_dir):
     return library
 
 def getWrappedRoutineNames(wrapped_routines_file):
-    fo = open(wrapped_routines_file)
```
This was never closed before

eric-wieser commented:
Also, the upside of us arguing about names is that by rebasing again, I remembered to fix the readme and comments to match the new names

charris commented:
In it goes. Thanks Eric.
